test_conformance/api/test_kernel_local_memory_size.cpp

```diff
@@ -0,0 +1,351 @@
+//
+// Copyright (c) 2020 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include <cinttypes>
+
+#include "testBase.h"
+#include "harness/testHarness.h"
+#include <memory>
+
+static const char* empty_kernel = R"CLC(
+__kernel void empty_kernel() {
+}
+)CLC";
+
+static const char* local_memory_kernel = R"CLC(
+__kernel void local_memory_kernel(global int* data) {
+    __local int array[10];
+
+    size_t id = get_global_id(0);
+    array[id] = 2 * id;
+    data[id] = array[id];
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+    if(id == 0)
+    {
+        for(size_t i = 0; i < 10; i++)
+            data[id] += array[i];
+    }
+}
+)CLC";
+
+static const char* local_param_kernel = R"CLC(
+__kernel void local_param_kernel(__local int* local_ptr, __global int* src,
+    __global int* dst) {
+
+    size_t id = get_global_id(0);
+
+    local_ptr[id] = src[id];
+    barrier(CLK_GLOBAL_MEM_FENCE);
+    dst[id] = local_ptr[id];
+    barrier(CLK_LOCAL_MEM_FENCE);
+    if(id == 9)
+    {
+        for(size_t i = 0; i < 10; i++)
+            dst[id] += local_ptr[i];
+    }
+}
+)CLC";
+
+static const char* local_param_local_memory_kernel = R"CLC(
+__kernel void local_param_local_memory_kernel(__local int* local_ptr,
+    __global int* src, __global int* dst) {
+
+    size_t id = get_global_id(0);
+
+    __local int local_data[10];
+    local_ptr[id] = src[id];
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+    if(id / 2 == 0) {
+        for(size_t i = 0; i < 10; i++)
+            local_data[id] += local_ptr[i];
+    }
+    else
+        local_data[id] = local_ptr[id] * 2;
```

Comment:
This isn't safe, for two reasons:

1. In the "if" branch, `local_data` may be read when it is uninitialized, due to `+=`.
1. There is no guarantee which of these branches are executed first, and `local_data` will have different results depending which executes first.  Theoretically, even, both branches could execute simultaneously, in which case there is a data race.

Really though, this kernel looks way too complicated for its purpose.  Can we do something simpler?